packages/runtime/datastore/src/dataStoreRuntime.ts

```diff
@@ -377,38 +391,52 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
         if (this.isAttached) {
             this.attachChannel(channel);
             return;
-        } else {
-            this.bind(channel.handle);
+        }
 
-            // If our data store is local then add the channel to the queue
-            if (!this.localChannelContextQueue.has(channel.id)) {
-                this.localChannelContextQueue.set(channel.id, this.contexts.get(channel.id) as LocalChannelContextBase);
-            }
+        /**
+         * If this channel has already been bound, do nothing. This can happen during attachGraph() when a channel's
+         * graph is attached. It calls bindToContext on the shared object which will end up back here.
+         */
+        if (this.boundhandles !== undefined && this.boundhandles.has(channel.handle)) {
+            return;
+        }
+
+        this.bind(channel.handle);
+
+        // If our data store is local then add the channel to the queue
+        if (!this.localChannelContextQueue.has(channel.id)) {
```

Comment:
It's rather hard to understand the meaning of this.localChannelContextQueue. And looking at isChannelAttached(), I think what would help is to remove notBoundedChannelContextSet / localChannelContextQueue / pendingAttach and instead have a property on channel that returns one of those states (enum). And have only one list of all channels (not IDs :), and do filters on top of it to get needed subset).

The other thought - I believe there is no need to have extra state (to track that enum), as channel already has that info (i.e. it's calculatable from existing state). I might be wrong, but the fact that we call various methods on channel (like attachChannel) suggests that channel has that info

Reply:
I agree that it's confusing but I would rather not do it as part of this PR. I can create an issue and do it separately as its mostly orthogonal to this change.

Reply:
Sounds good - please create an issue

Reply:
https://github.com/microsoft/FluidFramework/issues/9826